### DISCUSSION TITLE
ch6/ch6-06

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#163] [General]
# ch6/ch6-06

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch6/ch6-06.html

**Comment by zhangbest5** [comment]
这翻译的，真的是机翻，为绕而绕

**Reply by xm-xw** [reply]
是的  上上篇我也懒得看 还不如自己实战呢
我印象中go语言圣经没有这么拉的

**Comment by jiajialinli** [comment]
没有理解这篇的精神

**Comment by ccsuper1024** [comment]
这一节原文就写的这么烂，不知道为什么要写的这么复杂

**Comment by 8vd8** [comment]
其实感觉有些篇并没有编好，看进去要花时间，结果还发现一点用没有

**Comment by ice-jade** [comment]
“换句话说，相对于*s这个表达式会出现在所有的包中，s.words只需要在定义IntSet的包中出现（译注：所以还是推荐后者吧的意思）”
只要把类型名换成首字母小写不就解决了吗。这个例子举的不合适，反而让我觉得不用struct更好。

**Reply by soondubu137** [reply]
把类型名换成小写的话在别的package里就用不了了。写`intset`这个package的目的就是为了可以在其他package里使用`IntSet`，如果不export它的话这个package还有什么意义呢？如果不封装直接export的话用户就可以访问到这个slice，所以显然用struct封装是更好的做法。

**Comment by iriskl** [comment]
### 6.6 封装

封装是指对象的某些变量或方法对外部不可见，通常也称为信息隐藏。封装是面向对象编程的一个关键方面。

#### 封装的实现

- 在Go中，只有一种控制可见性的手段：大写字母开头的标识符是导出的（对外可见），小写字母开头的标识符是未导出的（对外不可见）。
- 这种控制方式适用于包内的所有成员，包括struct字段和方法。
- 例如，`IntSet`被定义为一个struct，以确保其字段`words`对外不可见：
    ```go
    type IntSet struct {
        words []uint64
    }
    ```

#### 封装的优点

1. **减少复杂度**：调用方不能直接修改对象的变量值，只需要关注少量的语句和变量的可能值。
    - 例如，`IntSet`的实现：
    ```go
    func (s *IntSet) Add(x int) {
        word, bit := x/64, uint(x%64)
        for word >= len(s.words) {
            s.words = append(s.words, 0)
        }
        s.words[word] |= 1 << bit
    }
    ```

2. **隐藏实现细节**：防止调用方依赖可能变化的具体实现，使设计者在不破坏API的情况下有更大的自由。
    - 例如，`bytes.Buffer`通过封装优化内存分配：
    ```go
    type Buffer struct {
        buf     []byte
        initial [64]byte
    }

    func (b *Buffer) Grow(n int) {
        if b.buf == nil {
            b.buf = b.initial[:0]
        }
        if len(b.buf)+n > cap(b.buf) {
            buf := make([]byte, b.Len(), 2*cap(b.buf) + n)
            copy(buf, b.buf)
            b.buf = buf
        }
    }
    ```

3. **保证数据一致性**：防止外部随意修改对象内部变量，只允许通过特定方法访问或修改。
    - 例如，`Counter`类型：
    ```go
    type Counter struct { n int }
    func (c *Counter) N() int     { return c.n }
    func (c *Counter) Increment() { c.n++ }
    func (c *Counter) Reset()     { c.n = 0 }
    ```

#### Getter和Setter

- 封装的对象通常提供getter和setter方法来访问和修改内部变量。
- 例如，`log.Logger`类型的getter和setter方法：
    ```go
    type Logger struct {
        flags  int
        prefix string
    }

    func (l *Logger) Flags() int
    func (l *Logger) SetFlags(flag int)
    func (l *Logger) Prefix() string
    func (l *Logger) SetPrefix(prefix string)
    ```

#### 封装的实例

- 封装有时并不是理想的选择，某些情况下需要暴露内部内容。
- 例如，`time.Duration`将其表现为`int64`的纳秒值，使得可以进行数值操作：
    ```go
    const day = 24 * time.Hour
    fmt.Println(day.Seconds()) // "86400"
    ```

- 对比`IntSet`和`geometry.Path`：
    - `Path`是一个坐标点的序列，可以直接暴露为slice。
    - `IntSet`可能需要额外字段来记录元素个数，因此采用封装。

#### 总结

封装在面向对象编程中非常重要，它不仅简化了代码，还隐藏了实现细节，保证了数据的一致性。在Go中，封装通过大写和小写字母来控制可见性。封装和方法是OOP编程的基础，下一章将介绍接口，进一步完善OOP编程。

**Comment by q1ngy** [comment]
实在看不下去...
✅ 本页的重点归纳
	•	Go 没有传统面向对象语言中的 private / public 关键字，而是用首字母大小写来控制包级可见性，从而实现封装。
	•	封装可以隐藏内部实现，让外部只能通过定义好的方法访问或操作，从而增强代码的健壮性和灵活性。
	•	在设计的时候要权衡封装与便利性 — 有时候暴露部分内部细节能带来使用上的便利。
